src/protocol/vm/state.rs

```diff
@@ -279,25 +286,34 @@ impl VMPoolState<PreCachedDB> {
 
         let sim_result = engine
             .simulate(&sim_params)
-            .map_err(TychoSimulationError::SimulationFailure)?;
+            .map_err(|err| TychoSimulationError::from(VMError::SimulationFailure(err)))?;
 
-        let address = decode(&[ParamType::Address], &sim_result.result)?
+        let address = decode(&[ParamType::Address], &sim_result.result)
+            .map_err(|_| {
+                TychoSimulationError::from(VMError::DecodingError("Failed to decode ABI".into()))
+            })?
             .into_iter()
             .next()
             .ok_or_else(|| {
-                TychoSimulationError::DecodingError("Expected an address in the result".into())
+                TychoSimulationError::from(VMError::DecodingError(
+                    "Expected an address in the result".into(),
+                ))
             })?;
 
         address
             .to_string()
             .parse()
-            .map_err(|_| TychoSimulationError::DecodingError("Expected an Address".into()))
+            .map_err(|_| {
+                TychoSimulationError::from(VMError::DecodingError("Expected an Address".into()))
```

Comment:
This error isn't a missing address, it's a badly encoded one. This will be more like `failed to decode` than 'expected a missing address'